heat/core/manipulations.py

```diff
@@ -1908,6 +1909,172 @@ def reshape_argsort_counts_displs(
 DNDarray.reshape.__doc__ = reshape.__doc__
 
 
+def roll(
+    x: DNDarray, shift: Union[int, Tuple[int]], axis: Optional[Union[int, Tuple[int]]] = None
+) -> DNDarray:
+    """
+    Rolls array elements along a specified axis. Array elements that roll beyond the last position are re-introduced at the first position.
+    Array elements that roll beyond the first position are re-introduced at the last position.
+
+    Parameters
+    ----------
+    x : DNDarray
+        input array
+    shift : Union[int, Tuple[int, ...]]
+        number of places by which the elements are shifted. If 'shift' is a tuple, then 'axis' must be a tuple of the same size, and each of
+        the given axes is shifted by the corrresponding element in 'shift'. If 'shift' is an `int` and 'axis' a `tuple`, then the same shift
+        is used for all specified axes.
+    axis : Optional[Union[int, Tuple[int, ...]]]
+        axis (or axes) along which elements to shift. If 'axis' is `None`, the array is flattened, shifted, and then restored to its original shape.
+        Default: `None`.
+
+    Raises
+    ------
+    TypeError
+        If 'shift' or 'axis' is not of type `int`, `list` or `tuple`.
+    ValueError
+        If 'shift' and 'axis' are tuples with different sizes.
+
+    Examples
+    --------
+    >>> a = ht.arange(20).reshape((4,5))
+    >>> a
+    DNDarray([[ 0,  1,  2,  3,  4],
+          [ 5,  6,  7,  8,  9],
+          [10, 11, 12, 13, 14],
+          [15, 16, 17, 18, 19]], dtype=ht.int32, device=cpu:0, split=None)
+    >>> ht.roll(a, 1)
+    DNDarray([[19,  0,  1,  2,  3],
+          [ 4,  5,  6,  7,  8],
+          [ 9, 10, 11, 12, 13],
+          [14, 15, 16, 17, 18]], dtype=ht.int32, device=cpu:0, split=None)
+    >>> ht.roll(a, -1, 0)
+    DNDarray([[ 5,  6,  7,  8,  9],
+          [10, 11, 12, 13, 14],
+          [15, 16, 17, 18, 19],
+          [ 0,  1,  2,  3,  4]], dtype=ht.int32, device=cpu:0, split=None)
+    """
+    sanitation.sanitize_in(x)
+
+    if axis is None:
+        return roll(x.flatten(), shift, 0).reshape(x.shape, new_split=x.split)
+
+    # inputs are ints
+    if isinstance(shift, int):
+        if isinstance(axis, int):
+            if x.split is not None and (axis == x.split or (axis + x.ndim) == x.split):
+                # roll along split axis
+                size = x.comm.Get_size()
+                rank = x.comm.Get_rank()
+
+                lshape_map = x.create_lshape_map()[:, x.split]  # local elements along axis
```

Comment:
might be able to save on some communication by turning off the `force_check` parameter in this function